# REST APIs

### REST APIs

* (**RE**presentational **S**tate **T**ransfer)
* Most popular way (paradigm) in which web services are structured nowaday
* And so are REST APIs.
* REST defines 6 architectural constraints#
* Those make any web service a RESTful API.
* You can find them [here](https://www.geeksforgeeks.org/rest-api-architectural-constraints/#:~:text=The%20only%20optional%20constraint%20of,API%20and%20Non-REST%20API.&text=For%20example%3A%20API%2Fusers.). (These concepts can be explained to more advanced classes, added as **optional** content)

* REST aims for a unified standard of interfaces
* This is important because it generalizes the use of **HTTP** (hypertext transfer protocol). how clients and servers make **requests** among each other.
* It is the tool we will be using when interacting with an API.
* Like a python object, HTTP has different methods, including the most important ones **GET** and **POST**. We will take a look at it in the next lesson.

In [1]:
# Import requests library
import requests

google = requests.get("https://developers.google.com")
print("Google:", google.status_code)

NBA =  requests.get("https://api.sportsdata.io/api/nba/fantasy/json/CurrentSeason")
print("NBA:", NBA.status_code)

rotten_tomato = requests.get("http://api.rottentomatoes.com/api/public/v1.0/lists/movies/box_office.json")
print("Rotten Tomatoes:", rotten_tomato.status_code)

Google: 200
NBA: 401
Rotten Tomatoes: 403


* `200`: Everything went ok
* `301`: The server is redirecting you to a different endpoint
* `400`: The server thinks you mad a bad request. You're not sending the right data for a request
* `401`: You're not properly authenticated
* `403`: Resource you're trying to access is forbidden
* `404`: Resource does not exist
* `503`: The server can't handle the request

# Use the skyscanner API from rapidapi

[link](https://rapidapi.com/skyscanner/api/skyscanner-flight-search) to the API 

In [ ]:
import getpass
# your rapid-api key here
api_key = getpass.getpass()

In [35]:
import requests

# send a GET request for locations 
url = "https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/autosuggest/v1.0/UK/GBP/en-GB/"

querystring = {"query":"Stockholm"}

headers = {
    'x-rapidapi-key': "",
    'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com"
    }

response = requests.request("GET", url, headers=headers, params=querystring)

print(response.text)

{
  "Places" : [ {
    "PlaceId" : "STOC-sky",
    "PlaceName" : "Stockholm",
    "CountryId" : "SE-sky",
    "RegionId" : "",
    "CityId" : "STOC-sky",
    "CountryName" : "Sweden"
  }, {
    "PlaceId" : "ARN-sky",
    "PlaceName" : "Stockholm Arlanda",
    "CountryId" : "SE-sky",
    "RegionId" : "",
    "CityId" : "STOC-sky",
    "CountryName" : "Sweden"
  }, {
    "PlaceId" : "NYO-sky",
    "PlaceName" : "Stockholm Skavsta",
    "CountryId" : "SE-sky",
    "RegionId" : "",
    "CityId" : "STOC-sky",
    "CountryName" : "Sweden"
  }, {
    "PlaceId" : "BMA-sky",
    "PlaceName" : "Stockholm Bromma",
    "CountryId" : "SE-sky",
    "RegionId" : "",
    "CityId" : "STOC-sky",
    "CountryName" : "Sweden"
  }, {
    "PlaceId" : "VST-sky",
    "PlaceName" : "Stockholm Vasteras",
    "CountryId" : "SE-sky",
    "RegionId" : "",
    "CityId" : "STOC-sky",
    "CountryName" : "Sweden"
  } ]
}


In [50]:
# send a GET request for location "Tokyo"
url = "https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/autosuggest/v1.0/UK/GBP/en-GB/"

params = {"query":"Tokyo"}

headers = {'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
        'x-rapidapi-key': str(api_key)}

response = requests.get(url, headers = headers, params = params)
response.json()

{'Places': [{'PlaceId': 'TYOA-sky',
   'PlaceName': 'Tokyo',
   'CountryId': 'JP-sky',
   'RegionId': '',
   'CityId': 'TYOA-sky',
   'CountryName': 'Japan'},
  {'PlaceId': 'NRT-sky',
   'PlaceName': 'Tokyo Narita',
   'CountryId': 'JP-sky',
   'RegionId': '',
   'CityId': 'TYOA-sky',
   'CountryName': 'Japan'},
  {'PlaceId': 'HND-sky',
   'PlaceName': 'Tokyo Haneda',
   'CountryId': 'JP-sky',
   'RegionId': '',
   'CityId': 'TYOA-sky',
   'CountryName': 'Japan'},
  {'PlaceId': 'TJH-sky',
   'PlaceName': 'Toyooka',
   'CountryId': 'JP-sky',
   'RegionId': '',
   'CityId': 'JTJH-sky',
   'CountryName': 'Japan'},
  {'PlaceId': 'OOK-sky',
   'PlaceName': 'Toksook Bay',
   'CountryId': 'US-sky',
   'RegionId': 'AK',
   'CityId': 'OOKA-sky',
   'CountryName': 'United States'},
  {'PlaceId': 'TKZ-sky',
   'PlaceName': 'Tokoroa',
   'CountryId': 'NZ-sky',
   'RegionId': '',
   'CityId': 'TKZN-sky',
   'CountryName': 'New Zealand'}]}

In [66]:
# send a request for flight quotes from San Francisco to New York

url = "https://skyscanner-skyscanner-flight-search-v1.p.rapidapi.com/apiservices/browsequotes/v1.0/US/USD/en-US/SFO-sky/NYCA-sky/anytime"

params = {"inboundpartialdate":"2021-12-12"}

headers = {
    'x-rapidapi-host': "skyscanner-skyscanner-flight-search-v1.p.rapidapi.com",
    'x-rapidapi-key': str(api_key)}

response = requests.get(url, headers=headers, params=params)

response.json()

{'Quotes': [{'QuoteId': 1,
   'MinPrice': 96,
   'Direct': False,
   'OutboundLeg': {'CarrierIds': [1065],
    'OriginId': 81727,
    'DestinationId': 65633,
    'DepartureDate': '2021-05-18T00:00:00'},
   'QuoteDateTime': '2021-05-01T08:59:00'},
  {'QuoteId': 2,
   'MinPrice': 118,
   'Direct': True,
   'OutboundLeg': {'CarrierIds': [851],
    'OriginId': 81727,
    'DestinationId': 60987,
    'DepartureDate': '2021-10-19T00:00:00'},
   'QuoteDateTime': '2021-04-30T14:56:00'},
  {'QuoteId': 3,
   'MinPrice': 118,
   'Direct': True,
   'OutboundLeg': {'CarrierIds': [851],
    'OriginId': 81727,
    'DestinationId': 60987,
    'DepartureDate': '2021-12-03T00:00:00'},
   'QuoteDateTime': '2021-05-04T11:48:00'},
  {'QuoteId': 4,
   'MinPrice': 119,
   'Direct': True,
   'OutboundLeg': {'CarrierIds': [851],
    'OriginId': 81727,
    'DestinationId': 60987,
    'DepartureDate': '2021-11-01T00:00:00'},
   'QuoteDateTime': '2021-05-03T07:12:00'},
  {'QuoteId': 5,
   'MinPrice': 119,
   'Dire

In [67]:
# Import libraries
import pandas as pd
import json

# Normalize the response
pd.json_normalize(response.json())

Quotes  \
0  [{'QuoteId': 1, 'MinPrice': 96, 'Direct': Fals...   

                                            Carriers  \
0  [{'CarrierId': 1065, 'Name': 'Frontier Airline...   

                                              Places  \
0  [{'Name': 'New York Newark', 'Type': 'Station'...   

                                          Currencies  
0  [{'Code': 'USD', 'Symbol': '$', 'ThousandsSepa...

In [71]:
quotes = pd.DataFrame(pd.json_normalize(response.json())['Quotes'][0])
carriers = pd.DataFrame(pd.json_normalize(response.json())['Carriers'][0])
places = pd.DataFrame(pd.json_normalize(response.json())['Places'][0])
currencies = pd.DataFrame(pd.json_normalize(response.json())['Currencies'][0])

In [73]:
quotes.head()

QuoteId  MinPrice  Direct  \
0        1        96   False   
1        2       118    True   
2        3       118    True   
3        4       119    True   
4        5       119    True   

                                         OutboundLeg        QuoteDateTime  
0  {'CarrierIds': [1065], 'OriginId': 81727, 'Des...  2021-05-01T08:59:00  
1  {'CarrierIds': [851], 'OriginId': 81727, 'Dest...  2021-04-30T14:56:00  
2  {'CarrierIds': [851], 'OriginId': 81727, 'Dest...  2021-05-04T11:48:00  
3  {'CarrierIds': [851], 'OriginId': 81727, 'Dest...  2021-05-03T07:12:00  
4  {'CarrierIds': [851], 'OriginId': 81727, 'Dest...  2021-05-03T12:41:00

In [74]:
carriers.head()

CarrierId               Name
0       1065  Frontier Airlines
1        851    Alaska Airlines
2       1793             United
3        870            jetBlue

In [75]:
places.head()

Name     Type  PlaceId IataCode SkyscannerCode  \
0              New York Newark  Station    50290      EWR            EWR   
1     New York John F. Kennedy  Station    60987      JFK            JFK   
2           New York LaGuardia  Station    65633      LGA            LGA   
3  San Francisco International  Station    81727      SFO            SFO   

        CityName CityId    CountryName  
0       New York   NYCA  United States  
1       New York   NYCA  United States  
2       New York   NYCA  United States  
3  San Francisco   SFOA  United States

In [76]:
currencies

Code Symbol ThousandsSeparator DecimalSeparator  SymbolOnLeft  \
0  USD      $                  ,                .          True   

   SpaceBetweenAmountAndSymbol  RoundingCoefficient  DecimalDigits  
0                        False                    0              2

Looks like a relational database packed into this JSON with a primary / foreign key structure!